flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/expressions/AggregateVisitors.java

```diff
@@ -0,0 +1,178 @@
+
+	private static final Map<FunctionDefinition, SqlAggFunction> AGG_DEF_SQL_OPERATOR_MAPPING = new HashMap<>();
+
+	static {
+		AGG_DEF_SQL_OPERATOR_MAPPING.put(BuiltInFunctionDefinitions.AVG, FlinkSqlOperatorTable.AVG);
+		AGG_DEF_SQL_OPERATOR_MAPPING.put(BuiltInFunctionDefinitions.COUNT, FlinkSqlOperatorTable.COUNT);
+		AGG_DEF_SQL_OPERATOR_MAPPING.put(BuiltInFunctionDefinitions.MAX, FlinkSqlOperatorTable.MAX);
+		AGG_DEF_SQL_OPERATOR_MAPPING.put(BuiltInFunctionDefinitions.MIN, FlinkSqlOperatorTable.MIN);
+		AGG_DEF_SQL_OPERATOR_MAPPING.put(BuiltInFunctionDefinitions.SUM, FlinkSqlOperatorTable.SUM);
+		AGG_DEF_SQL_OPERATOR_MAPPING.put(BuiltInFunctionDefinitions.SUM0, FlinkSqlOperatorTable.SUM0);
+		AGG_DEF_SQL_OPERATOR_MAPPING.put(BuiltInFunctionDefinitions.STDDEV_POP, FlinkSqlOperatorTable.STDDEV_POP);
+		AGG_DEF_SQL_OPERATOR_MAPPING.put(BuiltInFunctionDefinitions.STDDEV_SAMP, FlinkSqlOperatorTable.STDDEV_SAMP);
+		AGG_DEF_SQL_OPERATOR_MAPPING.put(BuiltInFunctionDefinitions.VAR_POP, FlinkSqlOperatorTable.VAR_POP);
+		AGG_DEF_SQL_OPERATOR_MAPPING.put(BuiltInFunctionDefinitions.VAR_SAMP, FlinkSqlOperatorTable.VAR_SAMP);
+		AGG_DEF_SQL_OPERATOR_MAPPING.put(BuiltInFunctionDefinitions.COLLECT, FlinkSqlOperatorTable.COLLECT);
+	}
+
+	static class AggFunctionVisitor extends ExpressionDefaultVisitor<SqlAggFunction> {
+		private final FlinkTypeFactory typeFactory;
+
+		AggFunctionVisitor(FlinkTypeFactory typeFactory) {
+			this.typeFactory = typeFactory;
+		}
+
+		@Override
+		public SqlAggFunction visit(CallExpression call) {
+			Preconditions.checkArgument(isFunctionOfKind(call, AGGREGATE));
+			FunctionDefinition def = call.getFunctionDefinition();
+			if (AGG_DEF_SQL_OPERATOR_MAPPING.containsKey(def)) {
+				return AGG_DEF_SQL_OPERATOR_MAPPING.get(def);
+			}
+			if (BuiltInFunctionDefinitions.DISTINCT.equals(def)) {
+				Expression innerAgg = call.getChildren().get(0);
+				return innerAgg.accept(this);
+			}
+			AggregateFunctionDefinition aggDef = (AggregateFunctionDefinition) def;
+			UserDefinedAggregateFunction userDefinedAggregateFunc = aggDef.getAggregateFunction();
+			if (userDefinedAggregateFunc instanceof AggregateFunction) {
+				AggregateFunction aggFunc = (AggregateFunction) userDefinedAggregateFunc;
+				return new AggSqlFunction(
+						aggFunc.functionIdentifier(),
+						aggFunc.toString(),
+						aggFunc,
+						fromLegacyInfoToDataType(aggDef.getResultTypeInfo()),
+						fromLegacyInfoToDataType(aggDef.getAccumulatorTypeInfo()),
+						typeFactory,
+						aggFunc.requiresOver());
+			} else {
+				throw new UnsupportedOperationException("TableAggregateFunction is not supported yet!");
+			}
+		}
+
+		@Override
+		protected SqlAggFunction defaultMethod(Expression expression) {
+			throw new TableException("Unexpected expression: " + expression);
+		}
+	}
+
+	/**
+	 * The class to get AggCall of Aggregate Expression.
+	 */
+	public static class AggCallVisitor extends ExpressionDefaultVisitor<RelBuilder.AggCall> {
```

Comment:
Can it use `AggFunctionVisitor` to construct AggCall? I see a lot of codes are repeated.

Reply:
Good suggestion!